rfcs/bring-ts.md

```diff
@@ -0,0 +1,248 @@
+$ ts2wing --version
+0.1.2
+```
+
+Let's say I want to use the [colors](https://www.npmjs.com/package/colors) library, first, let's
+import it using the `ts2wing` command:
+
+```sh
+$ ts2wing colors
+Installing colors@1.4.0 from npm and all dependencies into vendor/colors.wlib.src...
+Processing type definitions (index.d.ts)...
+Generating Wing wrapper classes...
+Compiling with JSII...
+Generating API.md...
+Packaging Wing Library to vendor/colors.wlib
+Cleaning vendor/colors.wlib.src
+```
+
+### The .wlib File
+
+This command created a Wing Library file under `vendor/colors.wlib`.
+
+> A `.wlib` file it a self-contained artifact that represents a reusable library in Wing. It
+> includes all the dependencies needed by the library (vendored in). (see note below about future
+> support for [public dependencies](#public-dependencies-are-not-supported)).
+
+Let's take a look at the API of our new library:
+
+```sh
+$ wing docs colors
+Opening browser...
+```
+
+This will open your browser with the generated API documentation for the generated library. Cute,
+ha?
+
+### Consuming the library
+
+OK, let's use this library in our Wing code.
+
+```js
+bring colors;
+
+print("${colors.Global.green("hello")}, ${colors.Colors.blue("world")}!")
+```
+
+As you can see, since Wing doesn't support global functions, the `Global` wrapper class was
+generated and includes a static method for each public API exported from `colors`.
+
+Similarly to turning global functions to static methods, `ts2wing` applies other heuristics to
+create an [adaptation layer](#type-system-adapters) between the type systems.
+
+### Escape hatch
+
+As you might have expected, there could be cases where `ts2wing` won't be able to create an adapter
+for a certain TypeScript library. It will take us time to establish all the patterns and in parallel
+evolve the Wing type system.
+
+If you run `ts2wing` with the switch `--unpacked` like so:
+
+```sh
+$ ts2wing edit colors
+OK, "colors" is ready to be edited under "vendor/colors.wlib.src".
+After you are done, make sure to run: ts2wing colors
+```
+
+Now, go ahead and make any modifications you like to `vendor/colors.wlib.src` and after
+you are finished, run:
+
+```sh
+$ ts2wing colors
+Found vendor/colors.wlib.src
+Processing type definitions (index.d.ts)...
+Generating Wing wrapper classes...
+Compiling with JSII...
+Generating API.md...
+Packaging Wing Library to vendor/colors.wlib
+Cleaning vendor/colors.wlib.src
+```
+
+This is it, now the updated `colors.wlib` will include your changes.
+
+### Type system adapters
+
+It is not possible to express every type in TypeScript within the Wing type system. This is because:
+
+1. Wing is young and it will take the type system time to evolve; 
+2. The design goals of Wing and TypeScript are different. TypeScript is designed to be able to
+   represent every possible type that can be expressed in JavaScript, while Wing is designed to be a
+   great language for cloud application programming.
```

Comment:
> I don't agree - it seems to me like JS's dynamic typing is very much a motivating reason why unions exist in TypeScript.

Theoretically, TS did not have to implement ADTs like this right? I figure it could be expressed more [akin to the way you discuss it here](https://github.com/winglang/wing/issues/977) while still supporting legacy APIs. 
I see your point though and will agree that what I said originally isn't fair, the reason they had to tackle the problem **is** that javascript allowed it. However, I think the particular solution they chose is a really good one when it may not have been the only option. 
Does that mean we have to do it the same way they did? No, but saying it's bad is not a good reason.

> Other languages address unions in a way that results in less compiler complexity (=> faster compile times):

You say less compiler complexity, I say less expressiveness. If we want that (which I suppose we do, based on design tenants) then that's fine, but we shouldn't pretend like what TS did is worse.

Reply:
I :heart: TypeScript's type system and I think it's beautiful and elegant and extremely powerful. And as we discussed, given the goals of the language, the type system is a work of art, it's inspiring and expressive.

But... as I mentioned, Wing doesn't have the same goals as TypeScript. And the decision whether to support type unions or not should be taken within the design goals and tradeoffs of Wing, not of TypeScript. And I don't think we need type unions in Wing. Like @Chriscbr says, most programming languages don't have support type unions. Yes, it might be less expressive for certain use cases, but there is also value in constrainting expressivity sometimes, especially in a language that aims to scale to large enterprises and huge multi-team projects.

I'll try to improve the reasoning, but this is not about aesthetics. The fact that we want to interop with TypeScript doesn't mean we need to carry over every design decision they made. Not because they are "wrong" for typescript, but because they might not be the right thing for Wing.

Reply:
> The decision whether to support type unions or not should be taken within the design goals and tradeoffs of Wing, not of TypeScript

Agreed, I think union types are a nice idea within the goals and tradeoffs of Wing regardless of TS interop or why they exist in the context of TS. We disagree on this, which is totally okay, but I wanted to make it clear that I don't feel that way because of the reasons you mention.

> The fact that we want to interop with TypeScript doesn't mean we need to carry over every design decision they made.

Then this RFC should strictly be about "What is the experience of using TS libs in Wing" rather than having notes about why we chose not to have certain features of the TS type system in Wing itself. If someone asks why we don't support unions here as-is, then the discussion should be moved elsewhere and focused on Wing's type system design so that it's free of any influence of TS interop.

Reply:
"there is also value in constrainting expressivity sometimes, especially in a language that aims to scale to large enterprises and huge multi-team projects." - I agree with @eladb here. 

Gonna use Golang as the example here, which by design is very simple and not as expressive as other languages e.g. JavaScript. But if you look at any tutorial/repo about Go, all the code looks the same! Thats very powerful because it allows new users to learn concepts very quickly without worrying that they are learning outdated syntax. 

If you take JS as an example for this you have the callbacks vs async/await. Prototypes vs class vs functional approach. This can be quite overwhelming for new users and even become a barrier to some hesitant on learning a new language.

I guess looping back to the design tenants it also fits in: 

- Syntax design aims to be concise and minimal, while being "batteries included" 

So i guess the question is do we see unions as being necessary for the wing ecosystem and not just something we added as a way to support another language?